combinator_logic_experiments/y_comb_examples.metta

```diff
@@ -0,0 +1,57 @@
+(= (I $x) $x)
+(= ((K $x) $y) $x)
+(= (K $x $y) $x)
+(= (S $x $y $z) ($x $z ($y $z)))
+
+(= (B $x $y $z) ($x ($y $z)))
+(= (C $x $y $z) ($x $y $z))
+(= (W $x $y) ($x $y $y))
+
+(= U (S I I))
+
+(= Y (S (K (S I I)) (S (S (K S) K) (K (S I I)))))
+
+(= Y (B U (C B U)))
+
+
+; want to check equivalence between SKI and BCKW systems
+!(assertEqual
+  (S (K S) K x y z)
+   (B x y z))
+
+!(assertEqual
+    (S (S (K (S (K S) K)) S) (K K) x y z)
+    (C x y z))
+
+!(assertEqual
+    (S S (S K) x y z)
+    (W x y z))
+
+; How to implement factorial throughout Y-combinator and compare two definitions of it???
+;(= (fact $x) (Y ( (lambda f)
+;		  (lambda $x) (if (> $x 0) (* $x (f (- x 1))) 1))))
+
+; to return function as an output of a function we can do stuff like that:
+(: Lambda (-> $a $t))
+(: lambda (-> Variable Atom Lambda))
```

Comment:
It looks like you wrote that `lambda` is a function that takes a variable, a body and returns a term of type `Lambda`, itself being a term of a function type, i.e. a function.  Did I read that right?

Are you sure that is what you want to express?  Wouldn't `lambda` be a function constructor instead?  Maybe you meant to say that `Lambda` is a function type as opposed to an inhabitant thereof?

Could you clarify what you mean?

Reply:
> Did I read that right?
I think so. 

> Could you clarify what you mean?
I want to have mechanism to construct functions that return functions. I don't really know how to describe this effectively in metta. But that piece of code just did what I want for that particular case. It was just kind of a shot in the dark.

Reply:
@ngeiswei  you are right, `lambda` is a function constructor but naming here is misleading. I would suggest replace type of lambda by `(: lambda (-> Variable Atom (-> $a $t)))`